File: Source/Engine/Graphics/D3D12/D3D12Buffer.cpp
// Copyright © Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "D3D12Buffer.h"
#include "D3D12CommandBuffer.h"
#include "D3D12Graphics.h"

namespace Alimer
{
	D3D12Buffer::D3D12Buffer(D3D12Graphics& device_, const BufferCreateInfo& info, const void* initialData)
		: Buffer(info)
		, device(device_)
	{
		D3D12_RESOURCE_DESC resourceDesc = {};
		resourceDesc.Dimension = D3D12_RESOURCE_DIMENSION_BUFFER;
		resourceDesc.Alignment = D3D12_DEFAULT_RESOURCE_PLACEMENT_ALIGNMENT;
		resourceDesc.Width = size;
        if ((usage & BufferUsage::Uniform) != 0)
        {
            // Align the buffer size to multiples of the dynamic uniform buffer minimum size
            resourceDesc.Width = AlignUp(resourceDesc.Width, D3D12_CONSTANT_BUFFER_DATA_PLACEMENT_ALIGNMENT);
        }
		resourceDesc.Height = 1;
		resourceDesc.DepthOrArraySize = 1;
		resourceDesc.MipLevels = 1;
		resourceDesc.Format = DXGI_FORMAT_UNKNOWN;
		resourceDesc.SampleDesc.Count = 1;
		resourceDesc.SampleDesc.Quality = 0;
		resourceDesc.Layout = D3D12_TEXTURE_LAYOUT_ROW_MAJOR;
		resourceDesc.Flags = D3D12_RESOURCE_FLAG_NONE;
		if (!CheckBitsAny(usage, BufferUsage::ShaderRead))
		{
			resourceDesc.Flags |= D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE;
		}
		if (CheckBitsAny(usage, BufferUsage::ShaderWrite))
		{
			resourceDesc.Flags |= D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS;
		}

		D3D12MA::ALLOCATION_DESC allocDesc{};
		allocDesc.HeapType = D3D12_HEAP_TYPE_DEFAULT;

		switch (memoryUsage)
		{
		case MemoryUsage::CpuOnly:
		case MemoryUsage::CpuToGpu:
			allocDesc.HeapType = D3D12_HEAP_TYPE_UPLOAD;
			state = D3D12_RESOURCE_STATE_GENERIC_READ;
			fixedResourceState = true;
			break;

		case MemoryUsage::GpuToCpu:
			allocDesc.HeapType = D3D12_HEAP_TYPE_READBACK;
			state = D3D12_RESOURCE_STATE_COPY_DEST;
			fixedResourceState = true;
			resourceDesc.Flags |= D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE;
			break;

		default:
			allocDesc.HeapType = D3D12_HEAP_TYPE_DEFAULT;
			break;
		}

		HRESULT result = device.GetAllocator()->CreateResource(&allocDesc,
			&resourceDesc,
			state,
			nullptr,
			&allocation,
			IID_PPV_ARGS(&handle)
		);

		if (FAILED(result))
		{
			LOGE("Direct3D12: Could not create buffer: {}", std::to_string(result));
			return;
		}

        if (info.label != nullptr)
        {
            SetName(info.label);
        }

        device.GetHandle()->GetCopyableFootprints(&resourceDesc, 0, 1, 0, nullptr, nullptr, nullptr, &allocatedSize);

		if (memoryUsage != MemoryUsage::GpuOnly)
		{
			ThrowIfFailed(handle->Map(0, nullptr, reinterpret_cast<void**>(&mappedData)));
		}

		if (initialData != nullptr)
		{
			if (memoryUsage == MemoryUsage::GpuOnly)
			{
				D3D12UploadContext uploadContext = device.ResourceUploadBegin(resourceDesc.Width);
				memcpy(uploadContext.CPUAddress, initialData, size);
				
				uploadContext.commandList->CopyBufferRegion(handle, 0,
                    uploadContext.resource, uploadContext.resourceOffset, size);

				device.ResourceUploadEnd(uploadContext);
			}
			else
			{
				memcpy(mappedData, initialData, size);
			}
		}

		gpuVirtualAddress = handle->GetGPUVirtualAddress();
        OnCreated();
	}

	D3D12Buffer::~D3D12Buffer()
	{
		Destroy();
	}

	void D3D12Buffer::Destroy()
	{
		if (handle != nullptr
			&& allocation != nullptr)
		{
			Unmap();
			device.DeferDestroy(handle, allocation);
		}

        D3D12GpuResource::Destroy();
        OnDestroyed();
	}

	void D3D12Buffer::ApiSetName()
	{
		auto wideName = ToUtf16(name);
        handle->SetName(wideName.c_str());
	}

	uint8_t* D3D12Buffer::Map()
	{
		if (!mapped && !mappedData)
		{
			D3D12_RANGE readRange = { };
			ThrowIfFailed(handle->Map(0, &readRange, reinterpret_cast<void**>(&mappedData)));
			mapped = true;
		}

		return mappedData;
	}

	void D3D12Buffer::Unmap()
	{
		if (mapped)
		{
            handle->Unmap(0, nullptr);
			mappedData = nullptr;
			mapped = false;
		}
	}

	void D3D12Buffer::Update(const void* data, size_t size, size_t offset)
	{
		if (memoryUsage == MemoryUsage::GpuOnly)
		{
		}
		else
		{
			memcpy(mappedData, data, size);
		}
	}
}

File: src/Alimer/Graphics/D3D12/D3D12Buffer.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using System.Runtime.CompilerServices;
using TerraFX.Interop.DirectX;
using TerraFX.Interop.Windows;
using static TerraFX.Interop.DirectX.D3D12;
using static TerraFX.Interop.DirectX.D3D12_HEAP_TYPE;
using static TerraFX.Interop.DirectX.D3D12_RESOURCE_FLAGS;
using static TerraFX.Interop.DirectX.D3D12_RESOURCE_STATES;
using static TerraFX.Interop.DirectX.D3D12_BARRIER_LAYOUT;
using static TerraFX.Interop.Windows.Windows;
using static TerraFX.Interop.Windows.E;

namespace Alimer.Graphics.D3D12;

internal unsafe class D3D12Buffer : GraphicsBuffer, ID3D12GpuResource
{
    private readonly D3D12GraphicsDevice _device;
    private readonly ComPtr<ID3D12Resource> _handle;
    private readonly nint _allocation;
    public readonly void* pMappedData;

    public D3D12Buffer(D3D12GraphicsDevice device, in BufferDescription description, void* initialData)
        : base(description)
    {
        _device = device;

        ulong alignedSize = description.Size;
        if ((description.Usage & BufferUsage.Constant) != 0)
        {
            alignedSize = MathUtilities.AlignUp(alignedSize, D3D12_CONSTANT_BUFFER_DATA_PLACEMENT_ALIGNMENT);
        }

        D3D12_RESOURCE_FLAGS resourceFlags = D3D12_RESOURCE_FLAG_NONE;
        if ((description.Usage & BufferUsage.ShaderWrite) != 0)
        {
            resourceFlags |= D3D12_RESOURCE_FLAG_ALLOW_UNORDERED_ACCESS;
        }

        if ((description.Usage & BufferUsage.ShaderRead) == 0 &&
            (description.Usage & BufferUsage.RayTracing) == 0)
        {
            resourceFlags |= D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE;
        }

        D3D12_RESOURCE_DESC1 resourceDesc = D3D12_RESOURCE_DESC1.Buffer(alignedSize, resourceFlags);

        D3D12MA.ALLOCATION_DESC allocationDesc = new()
        {
            HeapType = D3D12_HEAP_TYPE_DEFAULT
        };

        // https://microsoft.github.io/DirectX-Specs/d3d/D3D12EnhancedBarriers.html
        // Buffers may only use D3D12_BARRIER_LAYOUT_UNDEFINED as an initial layout.
        D3D12_BARRIER_LAYOUT initialLayout = D3D12_BARRIER_LAYOUT_UNDEFINED;
        D3D12_RESOURCE_STATES initialStateLegacy = D3D12_RESOURCE_STATE_COMMON;

        if (description.CpuAccess == CpuAccessMode.Read)
        {
            allocationDesc.HeapType = D3D12_HEAP_TYPE_READBACK;
            resourceDesc.Flags |= D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE;

            initialStateLegacy = D3D12_RESOURCE_STATE_COPY_DEST;
            ImmutableState = true;
        }
        else if (description.CpuAccess == CpuAccessMode.Write)
        {
            allocationDesc.HeapType = D3D12_HEAP_TYPE_UPLOAD;
            initialStateLegacy = D3D12_RESOURCE_STATE_GENERIC_READ;
            ImmutableState = true;
        }

        HRESULT hr = E_FAIL;
        if (device.EnhancedBarriersSupported)
        {
            hr = D3D12MA.Allocator_CreateResource3(
                device.MemoryAllocator,
                &allocationDesc,
                &resourceDesc,
                initialLayout,
                null,
                0, null,
                out _allocation,
                __uuidof<ID3D12Resource>(), (void**)_handle.GetAddressOf()
            );
        }
        else
        {
            hr = D3D12MA.Allocator_CreateResource2(
                device.MemoryAllocator,
                &allocationDesc,
                &resourceDesc,
                initialStateLegacy,
                null,
                out _allocation,
                __uuidof<ID3D12Resource>(), (void**)_handle.GetAddressOf()
            );
        }

        if (hr.FAILED)
        {
            Log.Error("D3D12: Failed to create Buffer.");
            return;
        }

        if (!string.IsNullOrEmpty(description.Label))
        {
            OnLabelChanged(description.Label!);
        }

        ulong allocatedSize;
        if (device.EnhancedBarriersSupported)
        {
            device.Device8->GetCopyableFootprints1(&resourceDesc,
                0, 1, 0, null, null, null, &allocatedSize);
        }
        else
        {
            device.Device->GetCopyableFootprints((D3D12_RESOURCE_DESC*)&resourceDesc,
                0, 1, 0, null, null, null, &allocatedSize);
        }

        GpuAddress = _handle.Get()->GetGPUVirtualAddress();
        AllocatedSize = allocatedSize;

        if (description.CpuAccess == CpuAccessMode.Read)
        {
            void* mappedData;
            ThrowIfFailed(_handle.Get()->Map(0, null, &mappedData));
            pMappedData = mappedData;
        }
        else if (description.CpuAccess == CpuAccessMode.Write)
        {
            D3D12_RANGE readRange = default;
            void* mappedData;
            ThrowIfFailed(_handle.Get()->Map(0, &readRange, &mappedData));
            pMappedData = mappedData;
        }

        // Issue data copy on request
        if (initialData != null)
        {
            D3D12UploadContext context = default;
            void* mappedData = null;
            if (description.CpuAccess == CpuAccessMode.Write)
            {
                mappedData = pMappedData;
            }
            else
            {
                context = device.Allocate(description.Size);
                mappedData = context.UploadBuffer.pMappedData;
            }

            Unsafe.CopyBlockUnaligned(mappedData, initialData, (uint)description.Size);
            //std::memcpy(mappedData, initialData, desc.size);

            if (context.IsValid)
            {
                context.CommandList.Get()->CopyBufferRegion(
                    _handle.Get(),
                    0,
                    context.UploadBuffer!.Handle,
                    0,
                    description.Size
                );

                device.Submit(in context);
            }
        }
    }

    public D3D12Buffer(D3D12GraphicsDevice device, ID3D12Resource* existingHandle, in BufferDescription descriptor)
        : base(descriptor)
    {
        _device = device;
        _handle = existingHandle;
    }

    /// <summary>
    /// Finalizes an instance of the <see cref="D3D12Buffer" /> class.
    /// </summary>
    ~D3D12Buffer() => Dispose(disposing: false);

    /// <inheritdoc />
    public override GraphicsDevice Device => _device;

    public ID3D12Resource* Handle => _handle;
    public bool ImmutableState { get; }

    public ulong GpuAddress { get; }
    public ulong AllocatedSize { get; }

    /// <inheitdoc />
    protected internal override void Destroy()
    {
        _ = D3D12MA.Allocation_Release(_allocation);
        _handle.Dispose();
    }

    /// <inheritdoc />
    protected override void OnLabelChanged(string? newLabel)
    {
         _handle.Get()->SetName(newLabel);
    }

    /// <inheitdoc />
    protected override void SetDataUnsafe(void* dataPtr, int offsetInBytes)
    {
        Unsafe.CopyBlockUnaligned((byte*)pMappedData + offsetInBytes, dataPtr, (uint)Size);
    }

    /// <inheitdoc />
    protected override void GetDataUnsafe(void* destPtr, int offsetInBytes)
    {
        Unsafe.CopyBlockUnaligned(destPtr, (byte*)pMappedData + offsetInBytes, (uint)Size);
    }
}
